File: src/main/scala/com/com/simplilearn/bigdata/spark/PersistSparkLauncher.scala
package com.com.simplilearn.bigdata.spark

import org.apache.hadoop.hbase.TableName
import org.apache.hadoop.hbase.client.{Get, Put, Table}
import org.apache.hadoop.hbase.util.Bytes
import org.apache.spark.sql._
import org.apache.spark.sql.types._


object PersistSparkLauncher {

  private val SEPERATOR = "-"

  def main(args: Array[String]) {
    if (args.length != 3) {
      System.out.println("Please provide <input_path> <spark_master> <namenode_config>")
      System.exit(0)
    }
    val inputPath = args(0)
    val namenodes = args(2)
    val dataset = readFile(inputPath, readWithHeader(getSparkSession("spark-retail-transaction-analysis-hbase-monthly", args(1))))
    persistOrderCountByBrandMonthly(dataset, namenodes)
    persistRevenueByBrandMonthly(dataset, namenodes)
    persistOrderCountByCategoryMonthly(dataset, namenodes)
    persistRevenueByCategoryMonthly(dataset, namenodes)
  }

  def getSparkSession(appName: String, master: String) = {
    val sparkSession = SparkSession.builder.appName(appName).master(if (master.equalsIgnoreCase("local")) "local[*]"
    else master).getOrCreate
    System.out.println("Spark version " + sparkSession.version)
    sparkSession
  }

  private def getSparkContezxt(appName: String, master: String) = getSparkSession(appName, master).sparkContext

  def readWithHeader(sparkSession: SparkSession) = {
    val transactionSchema = StructType(Array(
      StructField("ProductCode", StringType, true),
      StructField("Description", StringType, true),
      StructField("Brand", StringType, true),
      StructField("Category", StringType, true),
      StructField("SubCategory", StringType, true),
      StructField("ActualPrice", DoubleType, true),
      StructField("Marginpercentage", IntegerType, true),
      StructField("Discountpercentage", IntegerType, true),
      StructField("Quantity", IntegerType, true),
      StructField("TransactionAmount", DoubleType, true),
      StructField("InvoiceDate", TimestampType, true),
      StructField("Mode", StringType, true),
      StructField("InvoiceNumber", IntegerType, true),
      StructField("CustomerId", StringType, true),
      StructField("Country", StringType, true)));
    sparkSession.read.option("header", true).schema(transactionSchema).option("mode", "DROPMALFORMED")
  }

  def print(key: String, values: Array[AnyRef], colNames: Array[AnyRef]): Unit = {
    val logBuilder = new StringBuilder
    logBuilder.append(key + " = \n")
    var i = 0
    for (i <- 0 to colNames.length - 1) {
      logBuilder.append(colNames(i)).append(", ");
    }
    logBuilder.append("\n")

    for (i <- 0 to values.length - 1) {
      logBuilder.append(values(i)).append(", \n");
    }
    logBuilder.append("\n")
    System.out.println(logBuilder.toString);
  }


  def readFile(path: String, dataFrameReader: DataFrameReader) = {
    System.out.println("Reading file " + path)
    val dataset = dataFrameReader.csv(path)
    System.out.println("Dataset Schema " + dataset.schema)
    System.out.println("Row Count" + dataset.count())
    dataset
  }


  /**
    * @param dataset
    */
  private def persistOrderCountByBrandMonthly(dataset: Dataset[Row], namenodes: String): Unit = {
    var modifiedDataset = dataset.withColumn("Month", UDFUtils.toMonth(dataset("InvoiceDate"))).withColumn("Year", UDFUtils.toYear(dataset("InvoiceDate")))
    modifiedDataset = modifiedDataset.select("Year", "Month", "Brand", "Quantity").groupBy("Year", "Month", "Brand").agg(functions.sum("Quantity").as("Quantity"))
    var i = 0
    for (i <- 0 to 11) {
      val orderCountByBrandMonthly = modifiedDataset.filter("month = " + i).withColumn("MonthName", UDFUtils.toMonthName(modifiedDataset("Month"))).drop("Month")
      System.out.println("Row Count" + modifiedDataset.count())
      orderCountByBrandMonthly.foreach((row: Row) => {
        var table:Table = null
        try {
          val connection = HbaseManager.hbaseConnection(namenodes)
          val rowKey = Bytes.toBytes(row.get(0).toString + SEPERATOR + row.get(3).toString + SEPERATOR + row.get(1).toString)
          val get:Get = new Get(rowKey)
          table = connection.getTable(TableName.valueOf("brand_order_count_monthly"))
          val result = table.get(get)
          if (result.size  == 0) {
            val put:Put = new Put(rowKey)
            put.addColumn(Bytes.toBytes("brand_stats"), Bytes.toBytes("data"), Bytes.toBytes(row.get(2).asInstanceOf[Long]))
            table.put(put)
          }
        } finally try
          table.close
        catch {
          case ex: Exception =>
            ex.printStackTrace()
        }
      })
    }
  }

  private def persistRevenueByBrandMonthly(dataset: Dataset[Row], namenodes: String): Unit = {
    var modifiedDataset = dataset.withColumn("Month", UDFUtils.toMonth(dataset("InvoiceDate"))).withColumn("Year", UDFUtils.toYear(dataset("InvoiceDate")))
    modifiedDataset = modifiedDataset.select("Year", "Month", "Brand", "TransactionAmount").groupBy("Year", "Month", "Brand").agg(functions.sum("TransactionAmount").as("TransactionAmount"))
    var i = 0
    for (i <- 0 to 11) {
      val revenueByBrandMonthly = modifiedDataset.filter("month = " + i).withColumn("MonthName", UDFUtils.toMonthName(modifiedDataset("Month"))).drop("Month")
      System.out.println("Row Count" + modifiedDataset.count())
      revenueByBrandMonthly.foreach((row: Row) => {
        var table:Table = null
        try {
          val connection = HbaseManager.hbaseConnection(namenodes)
          val rowKey = Bytes.toBytes(row.get(0).toString + SEPERATOR + row.get(3).toString + SEPERATOR + row.get(1).toString)
          val get:Get = new Get(rowKey)
          table = connection.getTable(TableName.valueOf("brand_total_revenue_monthly"))
          val result = table.get(get)
          if (result.size  == 0) {
            val put:Put = new Put(rowKey)
            put.addColumn(Bytes.toBytes("brand_stats"), Bytes.toBytes("data"), Bytes.toBytes(row.get(2).asInstanceOf[Double]))
            table.put(put)
          }
        } finally try
          table.close
        catch {
          case ex: Exception =>
            ex.printStackTrace()
        }
      })
    }
  }

  private def persistOrderCountByCategoryMonthly(dataset: Dataset[Row], namenodes: String): Unit = {
    var modifiedDataset = dataset.withColumn("Month", UDFUtils.toMonth(dataset("InvoiceDate"))).withColumn("Year", UDFUtils.toYear(dataset("InvoiceDate")))
    modifiedDataset = modifiedDataset.select("Year", "Month", "Category", "Quantity").groupBy("Year", "Month", "Category").agg(functions.sum("Quantity").as("Quantity"))
    var i = 0
    for (i <- 0 to 11) {
      val orderCountByCategoryMonthly = modifiedDataset.filter("month = " + i).withColumn("MonthName", UDFUtils.toMonthName(modifiedDataset("Month"))).drop("Month")
      System.out.println("Row Count" + modifiedDataset.count())
      orderCountByCategoryMonthly.foreach((row: Row) => {
        var table:Table = null
        try {
          val connection = HbaseManager.hbaseConnection(namenodes)
          val rowKey = Bytes.toBytes(row.get(0).toString + SEPERATOR + row.get(3).toString + SEPERATOR + row.get(1).toString)
          val get:Get = new Get(rowKey)
          table = connection.getTable(TableName.valueOf("category_order_count_monthly"))
          val result = table.get(get)
          if (result.size  == 0) {
            val put:Put = new Put(rowKey)
            put.addColumn(Bytes.toBytes("category_stats"), Bytes.toBytes("data"), Bytes.toBytes(row.get(2).asInstanceOf[Long]))
            table.put(put)
          }
        } finally try
          table.close
        catch {
          case ex: Exception =>
            ex.printStackTrace()
        }
      })
    }
  }

  /**
    * @param dataset
    * @param namenodes
    */
  private def persistRevenueByCategoryMonthly(dataset: Dataset[Row], namenodes: String): Unit = {
    var modifiedDataset = dataset.withColumn("Month", UDFUtils.toMonth(dataset("InvoiceDate"))).withColumn("Year", UDFUtils.toYear(dataset("InvoiceDate")))
    modifiedDataset = modifiedDataset.select("Year", "Month", "Category", "TransactionAmount").groupBy("Year", "Month", "Category").agg(functions.sum("TransactionAmount").as("TransactionAmount"))
    var i = 0
    for (i <- 0 to 11) {
      val revenueByCategoryMonthly = modifiedDataset.filter("month = " + i).withColumn("MonthName", UDFUtils.toMonthName(modifiedDataset("Month"))).drop("Month")
      System.out.println("Row Count" + modifiedDataset.count())
      revenueByCategoryMonthly.foreach((row: Row) => {
        var table:Table = null
        try {
          val connection = HbaseManager.hbaseConnection(namenodes)
          val rowKey = Bytes.toBytes(row.get(0).toString + SEPERATOR + row.get(3).toString + SEPERATOR + row.get(1).toString)
          val get:Get = new Get(rowKey)
          table = connection.getTable(TableName.valueOf("category_total_revenue_monthly"))
          val result = table.get(get)
          if (result.size  == 0) {
            val put:Put = new Put(rowKey)
            put.addColumn(Bytes.toBytes("category_stats"), Bytes.toBytes("data"), Bytes.toBytes(row.get(2).asInstanceOf[Double]))
            table.put(put)
          }
        } finally try
          table.close
        catch {
          case ex: Exception =>
            ex.printStackTrace()
        }
      })
    }
  }
}

File: spark-retail-transaction-analysis-master/src/main/java/com/simplilearn/bigdata/spark/PersistSparkLauncher.java
package com.simplilearn.bigdata.spark;

import org.apache.hadoop.hbase.TableName;
import org.apache.hadoop.hbase.client.*;
import org.apache.hadoop.hbase.util.Bytes;
import org.apache.spark.SparkContext;
import org.apache.spark.api.java.function.ForeachFunction;
import org.apache.spark.sql.*;
import org.apache.spark.sql.Row;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class PersistSparkLauncher {

    private static final Logger logger = LoggerFactory.getLogger(PersistSparkLauncher.class);
    private static final String SEPERATOR = "-";

    public static void main(String[] args) throws Exception{

        if (args.length != 3) {
            System.out.println("Please provide <input_path> <spark_master> <namenode_config>");
            System.exit(0);
        }

        String inputPath = args[0];
        String master = args[1];
        String namenodes = args[2];

        SparkSession sparkSession = getSparkSession("spark-retail-transaction-analysis-hbase-monthly", master);
        Dataset<Row> dataset = readFile(inputPath ,readWithHeader(sparkSession));

        Connection connection = HbaseManager.hbaseConnection(namenodes);
        persistOrderCountByBrandMonthly(dataset, namenodes);
        persistRevenueByBrandMonthly(dataset, namenodes);
        persistOrderCountByCategoryMonthly(dataset, namenodes);
        persistRevenueByCategoryMonthly(dataset, namenodes);

    }

    private static SparkSession getSparkSession(String appName, String master) {
        SparkSession sparkSession = SparkSession.builder()
                .appName(appName)
                .master(master.equalsIgnoreCase("local") ? "local[*]" : master)
                .getOrCreate();
        print(new String[] {"Spark version"}, new Object[] {sparkSession.version()});
        UDFUtils.registerUDFs(sparkSession);
        return sparkSession;
    }

    private static SparkContext getSparkContezxt(String appName, String master) {
        return getSparkSession(appName, master).sparkContext();
    }

    private static DataFrameReader readWithHeader(SparkSession sparkSession) {
        return sparkSession.read()
                .option("header",true)
                .option("inferSchema", true)
                .option("mode", "DROPMALFORMED");
    }

    private static void print(String[] key, Object[] values) {
        StringBuilder logBuilder = new StringBuilder();
        for(int i =0 ;i < key.length ; i++) {
            logBuilder.append(key[i] +" = "+values[i])
                    .append("\n");
        }
        System.out.println(logBuilder.toString());
    }

    private static void print(String key, Object[] values, Object[] colNames) {
        StringBuilder logBuilder = new StringBuilder();
        logBuilder.append(key +" = \n");
        for(int i =0 ;i < colNames.length ; i++) {
            logBuilder.append(colNames[i])
                    .append(", ");
        }
        logBuilder.append("\n");
        for(int i =0 ;i < values.length ; i++) {
            logBuilder.append(values[i])
                    .append("\n");
        }
        System.out.println(logBuilder.toString());
    }


    private static Dataset<Row> readFile(String path, DataFrameReader dataFrameReader) {
        print(new String[] {"Reading file "}, new Object[] {path});
        Dataset<Row> dataset = dataFrameReader.csv(path);
        print(new String[] {"Dataset Schema"}, new Object[] {dataset.schema()});
        print(new String[] {"Row Count"}, new Object[] {dataset.count()});
        return dataset;
    }

    /**
     * @param dataset
     */
    private static void persistOrderCountByBrandMonthly(Dataset<Row> dataset, final String namenodes) {
        dataset = dataset
                .withColumn("Month", functions.callUDF("toMonth", dataset.col("InvoiceDate")))
                .withColumn("Year", functions.callUDF("toYear", dataset.col("InvoiceDate")));
        dataset = dataset
                .select("Year", "Month", "Brand", "Quantity")
                .groupBy("Year", "Month", "Brand")
                .agg(functions.sum("Quantity").as("Quantity"));
        for(int i=0 ;i < 12; i++) {
            Dataset<Row> orderCountByBrandMonthly = dataset
                    .filter("month = "+i)
                    .withColumn("MonthName", functions.callUDF("toMonthName", dataset.col("Month"))).drop("Month");
            print(new String[] {"Row Count"}, new Object[] {orderCountByBrandMonthly.count()});

            orderCountByBrandMonthly.foreach((ForeachFunction<Row>) row -> {
                Table table = null;
                try {
                    Connection connection = HbaseManager.hbaseConnection(namenodes);
                    byte[] rowKey = Bytes.toBytes(row.get(0).toString() + SEPERATOR + row.get(3).toString() + SEPERATOR + row.get(1).toString());
                    Get get = new Get(rowKey);
                    table = connection.getTable(TableName.valueOf("brand_order_count_monthly"));
                    Result result = table.get(get);
                    if(result.size() == 0) {
                        Put put = new Put(rowKey);
                        put.addColumn(Bytes.toBytes("brand_stats"), Bytes.toBytes("data"), Bytes.toBytes((((Long)row.get(2))).toString()));
                        table.put(put);
                    }
                }finally {
                    try{
                        table.close();
                    }catch(Exception ex) {
                        ex.printStackTrace();
                    }
                }
            });
        }
    }

    /**
     * @param dataset
     */
    private static void persistRevenueByBrandMonthly(Dataset<Row> dataset, final String namenodes) {
        dataset = dataset
                .withColumn("Month", functions.callUDF("toMonth", dataset.col("InvoiceDate")))
                .withColumn("Year", functions.callUDF("toYear", dataset.col("InvoiceDate")));
        dataset = dataset
                .select("Year", "Month", "Brand", "TransactionAmount")
                .groupBy("Year", "Month", "Brand")
                .agg(functions.sum("TransactionAmount").as("TransactionAmount"));
        for(int i=0 ;i < 12; i++) {
            Dataset<Row> revenueByBrandMonthly = dataset
                    .filter("month = "+i)
                    .withColumn("MonthName", functions.callUDF("toMonthName", dataset.col("Month"))).drop("Month");
            print(new String[] {"Row Count"}, new Object[] {revenueByBrandMonthly.count()});

            revenueByBrandMonthly.foreach((ForeachFunction<Row>) row -> {
                Table table = null;
                try {
                    Connection connection = HbaseManager.hbaseConnection(namenodes);
                    byte[] rowKey = Bytes.toBytes(row.get(0).toString() + SEPERATOR + row.get(3).toString() + SEPERATOR + row.get(1).toString());
                    Get get = new Get(rowKey);
                    table = connection.getTable(TableName.valueOf("brand_total_revenue_monthly"));
                    Result result = table.get(get);
                    if(result.size() == 0) {
                        Put put = new Put(rowKey);
                        put.addColumn(Bytes.toBytes("brand_stats"), Bytes.toBytes("data"), Bytes.toBytes(((Double)row.get(2)).toString()));
                        table.put(put);
                    }
                }finally {
                    try{
                        table.close();
                    }catch(Exception ex) {
                        ex.printStackTrace();
                    }
                }
            });
        }
    }

    /**
     * @param dataset
     */
    private static void persistOrderCountByCategoryMonthly(Dataset<Row> dataset, final String namenodes) {
        dataset = dataset
                .withColumn("Month", functions.callUDF("toMonth", dataset.col("InvoiceDate")))
                .withColumn("Year", functions.callUDF("toYear", dataset.col("InvoiceDate")));
        dataset = dataset
                .select("Year", "Month", "Category", "Quantity")
                .groupBy("Year", "Month", "Category")
                .agg(functions.sum("Quantity").as("Quantity"));
        for(int i=0 ;i < 12; i++) {
            Dataset<Row> orderCountByCategoryMonthly = dataset
                    .filter("month = "+i)
                    .withColumn("MonthName", functions.callUDF("toMonthName", dataset.col("Month"))).drop("Month");
            print(new String[] {"Row Count"}, new Object[] {dataset.count()});
            orderCountByCategoryMonthly.foreach((ForeachFunction<Row>) row -> {
                Table table = null;
                try {
                    Connection connection = HbaseManager.hbaseConnection(namenodes);
                    byte[] rowKey = Bytes.toBytes(row.get(0).toString() + SEPERATOR + row.get(3).toString() + SEPERATOR + row.get(1).toString());
                    Get get = new Get(rowKey);
                    table = connection.getTable(TableName.valueOf("category_order_count_monthly"));
                    Result result = table.get(get);
                    if(result.size() == 0) {
                        Put put = new Put(rowKey);
                        put.addColumn(Bytes.toBytes("category_stats"), Bytes.toBytes("data"), Bytes.toBytes((((Long)row.get(2)).toString())));
                        table.put(put);
                    }
                }finally {
                    try{
                        table.close();
                    }catch(Exception ex) {
                        ex.printStackTrace();
                    }
                }
            });
        }
    }

    /**
     * @param dataset
     * @param namenodes
     */
    private static void persistRevenueByCategoryMonthly(Dataset<Row> dataset, final String namenodes) {
        dataset = dataset
                .withColumn("Month", functions.callUDF("toMonth", dataset.col("InvoiceDate")))
                .withColumn("Year", functions.callUDF("toYear", dataset.col("InvoiceDate")));
        dataset = dataset
                .select("Year", "Month", "Category", "TransactionAmount")
                .groupBy("Year", "Month", "Category")
                .agg(functions.sum("TransactionAmount").as("TransactionAmount"));
        for(int i=0 ;i < 12; i++) {
            Dataset<Row> revenueByCategoryMonthly = dataset
                    .filter("month = "+i)
                    .withColumn("MonthName", functions.callUDF("toMonthName", dataset.col("Month"))).drop("Month");
            print(new String[] {"Row Count"}, new Object[] {dataset.count()});
            revenueByCategoryMonthly.foreach((ForeachFunction<Row>) row -> {
                Table table = null;
                try {
                    Connection connection = HbaseManager.hbaseConnection(namenodes);
                    byte[] rowKey = Bytes.toBytes(row.get(0).toString() + SEPERATOR + row.get(3).toString() + SEPERATOR + row.get(1).toString());
                    Get get = new Get(rowKey);
                    table = connection.getTable(TableName.valueOf("category_total_revenue_monthly"));
                    Result result = table.get(get);
                    if(result.size() == 0) {
                        Put put = new Put(rowKey);
                        put.addColumn(Bytes.toBytes("category_stats"), Bytes.toBytes("data"), Bytes.toBytes(((Double)row.get(2)).toString()));
                        table.put(put);
                    }
                }finally {
                    try{
                        table.close();
                    }catch(Exception ex) {
                        ex.printStackTrace();
                    }
                }
            });
        }
    }
}
